be/src/exec/spill/input_stream.cpp

```diff
@@ -0,0 +1,242 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/spill/input_stream.h"
+
+#include "runtime/sorted_chunks_merger.h"
+#include "util/blocking_queue.hpp"
+#include "util/defer_op.h"
+
+namespace starrocks {
+namespace spill {
+
+static const int chunk_buffer_max_size = 2;
+
+class BufferedInputStream : public InputStream {
+public:
+    BufferedInputStream(int capacity, InputStreamPtr stream) : _capacity(capacity), _input_stream(stream) {}
+    ~BufferedInputStream() override = default;
+
+    bool is_buffer_full() { return _chunk_buffer.get_size() >= _capacity || eof(); }
+    bool has_chunk() { return !_chunk_buffer.empty() || eof(); }
+
+    StatusOr<ChunkUniquePtr> get_next(SerdeContext& ctx) override;
+    bool is_ready() override { return has_chunk(); }
+    void close() override {}
+
+    bool enable_prefetch() const override { return true; }
+
+    Status prefetch(SerdeContext& ctx) override;
+
+    StatusOr<ChunkUniquePtr> read_from_buffer();
+
+private:
+    bool _acquire() {
+        bool expected = false;
+        return _is_prefetching.compare_exchange_strong(expected, true);
+    }
+    void _release() { _is_prefetching = false; }
+
+private:
+    int _capacity;
+    InputStreamPtr _input_stream;
+    UnboundedBlockingQueue<ChunkUniquePtr> _chunk_buffer;
+    std::atomic_bool _is_prefetching = false;
+};
+
+StatusOr<ChunkUniquePtr> BufferedInputStream::read_from_buffer() {
+    if (_chunk_buffer.empty()) {
+        CHECK(eof());
+        return Status::EndOfFile("end of stream");
+    }
+    ChunkUniquePtr res;
+    CHECK(_chunk_buffer.try_get(&res));
+    return res;
+}
+
+StatusOr<ChunkUniquePtr> BufferedInputStream::get_next(SerdeContext& ctx) {
+    if (has_chunk()) {
+        return read_from_buffer();
+    }
+    CHECK(!_is_prefetching);
+    return _input_stream->get_next(ctx);
+}
+
+Status BufferedInputStream::prefetch(SerdeContext& ctx) {
+    if (is_buffer_full()) {
+        return Status::OK();
+    }
+    // concurrent prefetch is not allowed, should call _acquire and _release before and after prefetch
+    // to ensure that it doesn't happen.
+    if (!_acquire()) {
+        return Status::OK();
+    }
+    DeferOp defer([this]() { _release(); });
+
+    auto res = _input_stream->get_next(ctx);
+    if (res.ok()) {
+        _chunk_buffer.put(std::move(res.value()));
+        return Status::OK();
+    } else if (res.status().is_end_of_file()) {
+        mark_is_eof();
+        return Status::OK();
+    }
+    return res.status();
+}
+
+class UnorderedInputStream : public InputStream {
+public:
+    UnorderedInputStream(const std::vector<BlockPtr>& input_blocks, SerdePtr serde)
+            : InputStream(input_blocks), _serde(std::move(serde)) {}
+
+    ~UnorderedInputStream() = default;
+
+    StatusOr<ChunkUniquePtr> get_next(SerdeContext& ctx) override;
```

Comment:
could the ctx be const?

Reply:
> could the ctx be const?

we may read and write data to the buffer in SerdeContext, it can't be const